File: app/bot/sql_guard.py
import re


FORBIDDEN_KEYWORDS = {
    "alter",
    "call",
    "copy",
    "create",
    "delete",
    "do",
    "drop",
    "grant",
    "insert",
    "revoke",
    "truncate",
    "update",
}


def validate_select_query(sql: str) -> str:
    sql = strip_code_fence(sql).strip()
    sql = sql[:-1].strip() if sql.endswith(";") else sql

    if ";" in sql:
        raise ValueError("Multiple SQL statements are not allowed")

    lowered = sql.casefold()
    if not (lowered.startswith("select ") or lowered.startswith("with ")):
        raise ValueError("Only SELECT queries are allowed")

    tokens = set(re.findall(r"\b[a-z_]+\b", lowered))
    forbidden = tokens & FORBIDDEN_KEYWORDS
    if forbidden:
        raise ValueError(f"Forbidden SQL keyword: {sorted(forbidden)[0]}")

    if not re.search(r"\blimit\s+\d+\b", lowered):
        sql = f"{sql}\nLIMIT 50"

    return sql


def strip_code_fence(text: str) -> str:
    text = text.strip()
    if text.startswith("```"):
        lines = text.splitlines()
        if lines and lines[0].startswith("```"):
            lines = lines[1:]
        if lines and lines[-1].strip() == "```":
            lines = lines[:-1]
        return "\n".join(lines)
    return text

File: sql_guard.go
package main

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

var forbiddenKeywords = map[string]bool{
	"alter": true, "call": true, "copy": true, "create": true, "delete": true,
	"do": true, "drop": true, "grant": true, "insert": true, "revoke": true,
	"truncate": true, "update": true,
}

var sqlTokenPattern = regexp.MustCompile(`\b[a-z_]+\b`)
var limitPattern = regexp.MustCompile(`\blimit\s+\d+\b`)

func ValidateSelectQuery(query string) (string, error) {
	query = strings.TrimSpace(stripCodeFence(query))
	if strings.HasSuffix(query, ";") {
		query = strings.TrimSpace(strings.TrimSuffix(query, ";"))
	}
	if strings.Contains(query, ";") {
		return "", fmt.Errorf("multiple SQL statements are not allowed")
	}
	lowered := strings.ToLower(query)
	if !strings.HasPrefix(lowered, "select ") && !strings.HasPrefix(lowered, "with ") {
		return "", fmt.Errorf("only SELECT queries are allowed")
	}
	var forbidden []string
	for _, token := range sqlTokenPattern.FindAllString(lowered, -1) {
		if forbiddenKeywords[token] {
			forbidden = append(forbidden, token)
		}
	}
	if len(forbidden) > 0 {
		sort.Strings(forbidden)
		return "", fmt.Errorf("forbidden SQL keyword: %s", forbidden[0])
	}
	if !limitPattern.MatchString(lowered) {
		query += "\nLIMIT 50"
	}
	return query, nil
}

func stripCodeFence(text string) string {
	text = strings.TrimSpace(text)
	if !strings.HasPrefix(text, "```") {
		return text
	}
	lines := strings.Split(text, "\n")
	if len(lines) > 0 && strings.HasPrefix(lines[0], "```") {
		lines = lines[1:]
	}
	if len(lines) > 0 && strings.TrimSpace(lines[len(lines)-1]) == "```" {
		lines = lines[:len(lines)-1]
	}
	return strings.Join(lines, "\n")
}
